[CI] XContentParserTests#testSubParserArray fails frequently

Just saw three failures in the last few hours:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/2818/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+artifactory/80/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+artifactory/81/console

Reproduces locally with the following seed:
```
 ./gradlew :libs:x-content:unitTest \
  -Dtests.seed=7F3152E0775C4F75 \
  -Dtests.class=org.elasticsearch.common.xcontent.XContentParserTests \
  -Dtests.method="testSubParserArray" \
  -Dtests.security.manager=true \
  -Dtests.locale=es-VE \
  -Dtests.timezone=America/Virgin \
  -Dcompiler.java=12 \
  -Druntime.java=8
```

```
  > Throwable #1: java.lang.AssertionError: max must be >= min: 0, -1
   >    at __randomizedtesting.SeedInfo.seed([7F3152E0775C4F75:D0BA268FB44B084A]:0)
   >    at com.carrotsearch.randomizedtesting.generators.RandomNumbers.randomIntBetween(RandomNumbers.java:16)
   >    at com.carrotsearch.randomizedtesting.RandomizedTest.randomInt(RandomizedTest.java:164)
   >    at org.elasticsearch.test.ESTestCase.randomInt(ESTestCase.java:658)
   >    at org.elasticsearch.common.xcontent.XContentParserTests.testSubParserArray(XContentParserTests.java:393)
   >    at java.lang.Thread.run(Thread.java:748)
```

--- a/libs/x-content/src/test/java/org/elasticsearch/common/xcontent/XContentParserTests.java
+++ b/libs/x-content/src/test/java/org/elasticsearch/common/xcontent/XContentParserTests.java
@@ -20,7 +20,6 @@
 package org.elasticsearch.common.xcontent;
 
 import com.fasterxml.jackson.core.JsonParseException;
-
 import org.elasticsearch.common.CheckedSupplier;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.bytes.BytesReference;
@@ -369,7 +368,6 @@ public class XContentParserTests extends ESTestCase {
         }
     }
 
-    @AwaitsFix(bugUrl="https://github.com/elastic/elasticsearch/issues/40617")
     public void testSubParserArray() throws IOException {
         XContentBuilder builder = XContentFactory.jsonBuilder();
         int numberOfArrayElements = randomInt(10);
@@ -378,7 +376,7 @@ public class XContentParserTests extends ESTestCase {
         builder.startArray();
         int numberOfTokens = 0;
         for (int i = 0; i < numberOfArrayElements; ++i) {
-            numberOfTokens += generateRandomObjectForMarking(builder);
+            numberOfTokens += generateRandomObject(builder, 0);
         }
         builder.endArray();
         builder.endObject();
@@ -392,7 +390,7 @@ public class XContentParserTests extends ESTestCase {
             assertEquals(XContentParser.Token.START_ARRAY, parser.nextToken()); // [
             XContentParser subParser = new XContentSubParser(parser);
             try {
-                int tokensToSkip = randomInt(numberOfTokens - 1);
+                int tokensToSkip = randomInt(numberOfTokens);
                 for (int i = 0; i < tokensToSkip; i++) {
                     // Simulate incomplete parsing
                     assertNotNull(subParser.nextToken());

